{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from anndata import read_h5ad\n",
    "\n",
    "import sys\n",
    "import scanpy as sc\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import nbinom, pearsonr\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras import backend as K\n",
    "from tensorflow.keras.models import load_model\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "import pickle\n",
    "import anndata as ad\n",
    "from scipy.stats import zscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import sys\n",
    "# sys.path.insert(0, '/scratch/imb/Xiao/STimage')\n",
    "from stimage._utils import gene_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "def plot_correlation(df, attr_1, attr_2):\n",
    "    r = stats.pearsonr(df[attr_1], \n",
    "                       df[attr_2])[0] **2\n",
    "\n",
    "    g = sns.lmplot(data=df,\n",
    "        x=attr_1, y=attr_2,\n",
    "        height=5, legend=True\n",
    "    )\n",
    "    # g.set(ylim=(0, 360), xlim=(0,360))\n",
    "\n",
    "    g.set_axis_labels(attr_1, attr_2)\n",
    "    plt.annotate(r'$R^2:{0:.2f}$'.format(r),\n",
    "                (max(df[attr_1])*0.9, max(df[attr_2])*0.9))\n",
    "    return g\n",
    "\n",
    "\n",
    "def calculate_correlation(attr_1, attr_2):\n",
    "    r = stats.pearsonr(attr_1, \n",
    "                       attr_2)[0]\n",
    "    return r\n",
    "\n",
    "def calculate_correlation_2(attr_1, attr_2):\n",
    "    r = stats.spearmanr(attr_1, \n",
    "                       attr_2)[0]\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_PRED = Path(\"/scratch/project_mnt/S0010/Xiao/Q1851/Xiao/Wiener_backup/STimage_exp/stimage_LOOCV_skin_cancer_selected_gene\")\n",
    "PATH_GC = Path(\"/scratch/project_mnt/S0010/Xiao/Q2051/STimage_project/STimage_dataset/PROCESSED/SkinVisium\")\n",
    "OUT_PATH = PATH_PRED / \"PLOT\"\n",
    "OUT_PATH.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_adata = ad.read_h5ad(PATH_GC / \"all_adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_to_select = [\"Visium29_B1\", \"Visium29_C1\",  \"Visium38_B1\", \"Visium38_D1\", \"Visium37_D1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open pickle file\n",
    "with open(PATH_PRED / \"top_100gene.pickle\", \"rb\") as f:\n",
    "    top_100gene = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HSPB1\n",
      "B2M\n",
      "PFN1\n",
      "CALML5\n",
      "CD74\n",
      "CD44\n",
      "S100A10\n",
      "ANXA2\n",
      "IFI27\n",
      "CD63\n",
      "APOE\n",
      "ILF3\n",
      "FLNA\n",
      "CSTA\n",
      "RHOA\n",
      "S100B\n",
      "MYH9\n",
      "MYC\n",
      "CTNNB1\n",
      "CTSD\n",
      "KLF4\n",
      "CD81\n",
      "LGALS1\n",
      "MIF\n",
      "CTSB\n",
      "HLA-DRA\n",
      "HMGB1\n",
      "HLA-E\n",
      "SNCA\n",
      "PLA2G4F\n",
      "HMGA1\n",
      "TNC\n",
      "EIF4A2\n",
      "RXRA\n",
      "ZFP36L1\n",
      "S100A8\n",
      "CD99\n",
      "TAPBP\n",
      "GNAS\n",
      "RBX1\n",
      "ZFP36L2\n",
      "BCL2L2\n",
      "HLA-DMA\n",
      "S100A9\n",
      "FGFR3\n",
      "IQGAP1\n",
      "RAC1\n",
      "PRAME\n",
      "DDX3X\n",
      "CEBPA\n",
      "HLA-DPB1\n",
      "SDC1\n",
      "HLA-DPA1\n",
      "CDH1\n",
      "ERBB3\n",
      "ITGB1\n",
      "PPP2CB\n",
      "S100A7\n",
      "CTNND1\n",
      "SPINK5\n",
      "CCL19\n",
      "EMP3\n",
      "CDK4\n",
      "ANXA5\n",
      "MCL1\n",
      "MAP2K2\n",
      "GTF2I\n",
      "SOX10\n",
      "MDM4\n",
      "EPHB6\n",
      "TAP1\n",
      "PTPRC\n",
      "SERPINB2\n",
      "IFITM2\n",
      "IFI16\n",
      "LAMP1\n",
      "IL20RB\n",
      "RORA\n",
      "IKBKB\n",
      "AFDN\n",
      "PTEN\n",
      "MACF1\n",
      "LGALS3\n",
      "HRAS\n",
      "IRF4\n",
      "GATA3\n",
      "GNAI2\n",
      "C1R\n",
      "STAT6\n",
      "TOLLIP\n",
      "PSMD2\n",
      "NFE2L2\n",
      "MARCKSL1\n",
      "FN1\n",
      "MAF\n",
      "HLA-F\n",
      "KLF5\n",
      "CXCL14\n",
      "ARG1\n",
      "TFDP1\n"
     ]
    }
   ],
   "source": [
    "for i in top_100gene:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "skin_gene_list = [\n",
    "    \"CD63\",\n",
    "    \"CD74\",\n",
    "    \"LGALS1\",\n",
    "    \"SERPINB2\",\n",
    "    \"ANXA5\",\n",
    "    \"HLA-DRA\",\n",
    "    \"PFN1\",\n",
    "    \"B2M\",\n",
    "    \"CD44\",\n",
    "    \"S100A10\",\n",
    "    \"CTSB\",\n",
    "    \"HLA-DPA1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 B2M\n",
      "2 PFN1\n",
      "4 CD74\n",
      "5 CD44\n",
      "6 S100A10\n",
      "9 CD63\n",
      "22 LGALS1\n",
      "24 CTSB\n",
      "25 HLA-DRA\n",
      "52 HLA-DPA1\n",
      "63 ANXA5\n",
      "72 SERPINB2\n"
     ]
    }
   ],
   "source": [
    "for i, gene in enumerate(top_100gene):\n",
    "    if gene in skin_gene_list:\n",
    "        print(i, gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_plot = [\"S100B\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mSignature:\u001b[0m\n",
      "\u001b[0mgene_plot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0madata\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0manndata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_core\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0manndata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAnnData\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mmethod\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'CumSum'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mslot\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mgenes\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mUnion\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNoneType\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mthreshold\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfloat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mlibrary_id\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdata_alpha\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfloat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mtissue_alpha\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfloat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1.0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mvmin\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfloat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mvmax\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mfloat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mcmap\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'Spectral_r'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mspot_size\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mUnion\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mfloat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m6.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mshow_legend\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mshow_color_bar\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mshow_axis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mcropped\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mimage_scale\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mmargin\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0moutput\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mUnion\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmatplotlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbackends\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbackend_agg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFigureCanvasAgg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNoneType\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDocstring:\u001b[0m\n",
      "Gene expression plot for spatial transcriptomics data.\n",
      "\n",
      "Parameters\n",
      "----------\n",
      "adata : AnnData\n",
      "    Annotated data matrix.\n",
      "method : str, optional\n",
      "    Use method to count. We provide: NaiveMean, NaiveSum, CumSum.\n",
      "slot : str, optional\n",
      "    Slot name in adata.obsm.\n",
      "genes : str or list, optional\n",
      "    Choose a gene or a list of genes.\n",
      "threshold : float, optional\n",
      "    Threshold to filter genes.\n",
      "library_id : str, optional\n",
      "    Library id stored in AnnData.\n",
      "data_alpha : float, optional\n",
      "    Opacity of the spot.\n",
      "tissue_alpha : float, optional\n",
      "    Opacity of the tissue.\n",
      "vmin : float, optional\n",
      "    Minimum value for the color map.\n",
      "vmax : float, optional\n",
      "    Maximum value for the color map.\n",
      "cmap : str, optional\n",
      "    Color map to use.\n",
      "spot_size : float or int, optional\n",
      "    Size of the spot.\n",
      "show_legend : bool, optional\n",
      "    Show legend or not.\n",
      "show_color_bar : bool, optional\n",
      "    Show color bar or not.\n",
      "show_axis : bool, optional\n",
      "    Show axis or not.\n",
      "cropped : bool, optional\n",
      "    Crop the image or not.\n",
      "image_scale : bool, optional\n",
      "    Scale the image or not.\n",
      "margin : int, optional\n",
      "    Margin size.\n",
      "name : str, optional\n",
      "    Name of the output figure file.\n",
      "output : str, optional\n",
      "    Save the figure as file or not.\n",
      "\n",
      "Returns\n",
      "-------\n",
      "Optional[FigureCanvasAgg]\n",
      "    The generated gene expression plot as a FigureCanvasAgg object.\n",
      "\u001b[0;31mFile:\u001b[0m      ~/STimage/stimage/_utils.py\n",
      "\u001b[0;31mType:\u001b[0m      function"
     ]
    }
   ],
   "source": [
    "?gene_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in PATH_PRED.glob(\"./pred*\"):\n",
    "    sample = i.stem.split(\"pred_\")[1]\n",
    "    if sample in samples_to_select:\n",
    "        adata = ad.read_h5ad(i)\n",
    "        adata.X = adata.to_df().apply(zscore)\n",
    "        for gene in gene_to_plot:\n",
    "            try:\n",
    "                gene_plot(adata, genes=gene, image_scale=10, library_id=sample, spot_size= 30,\n",
    "                        output=str(OUT_PATH), name=f\"pred_{sample}_{gene}.pdf\")\n",
    "            except:\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples_to_select:\n",
    "    adata = all_adata[all_adata.obs[\"library_id\"] == sample]\n",
    "    # adata.X = adata.to_df().apply(zscore)\n",
    "    for gene in gene_to_plot:\n",
    "        try:\n",
    "            gene_plot(adata, genes=gene, image_scale=10, library_id=sample, spot_size= 30,\n",
    "                     output=str(OUT_PATH), name=f\"gc_{sample}_{gene}.pdf\")\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "stimage_ld",
   "language": "python",
   "name": "stimage_ld"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
